src/components/lava-coder/use-camera-controls.ts

```diff
@@ -1,21 +1,89 @@
-import { Cartesian3, Cartographic, CesiumWidget, Math as CSMath, sampleTerrainMostDetailed } from "@cesium/engine";
-import { useEffect } from "react";
+import {
+  Cartesian2, Cartesian3, Cartographic, CesiumWidget, Math as CSMath, HeadingPitchRange, sampleTerrainMostDetailed
+} from "@cesium/engine";
+import { useCallback, useEffect, useState } from "react";
+import { IOnDragArgs, useCesiumDragEvents } from "./use-cesium-drag-events";
+
+export const kDefaultCameraMode = "panning";
 
 const kDefaultXLng = -155.470;
 const kDefaultYLat = 19.150;
-const kMinDistanceAboveTerrain = 500;
-const kMaxDistanceAboveTerrain = 130000;
+const kMinDistanceAboveTerrain = 1000;
+const kMaxDistanceAboveTerrain = 135000;
+
+function getAngleFromCenter(pos: Cartesian2, center: Cartesian2) {
+  return Math.atan2(pos.y - center.y, pos.x - center.x);
+}
 
 export function useCameraControls(viewer: CesiumWidget | null, verticalExaggeration: number) {
 
-  function cameraMoveRate() {
+  const [cameraMode, setCameraMode] = useState<"pitch" | "heading" | "panning">(kDefaultCameraMode);
+
+  const handlePan = useCallback(({ dx, dy }: IOnDragArgs) => {
+    if (!viewer) return;
+
+    const { camera } = viewer;
+
+    // Adjust the sensitivity as needed
+    const panFactor = 0.5;
+
+    // Pan the camera: moveRight and moveUp are in meters, so you may want to scale by camera height
+    const moveRate = camera.positionCartographic.height / 500 * panFactor;
+    camera.moveRight(-dx * moveRate);
+    camera.moveUp(dy * moveRate);
+  }, [viewer]);
+
+  const handleRotateHeading = useCallback(({ startPosition, position, initialCameraState }: IOnDragArgs) => {
+    if (!viewer) return;
+
+    const { camera } = viewer;
+    const { center, target, initialHeading, initialPitch, initialRange } = initialCameraState;
+
+    const prevAngle = getAngleFromCenter(startPosition, center);
+    const currAngle = getAngleFromCenter(position, center);
+    const deltaTheta = currAngle - prevAngle; // Radians, positive is CCW
+
+    camera.lookAt(target, new HeadingPitchRange(
+      initialHeading - deltaTheta,
+      initialPitch,
+      initialRange
+    ));
+  }, [viewer]);
+
+  const handleRotatePitch = useCallback(({ dyTotal, initialCameraState }: IOnDragArgs) => {
+    if (!viewer) return;
+
+    const { camera } = viewer;
+    const { target, initialHeading, initialPitch, initialRange } = initialCameraState;
+
+    const pitchFactor = 0.01;
+
+    camera.lookAt(target, new HeadingPitchRange(
+      initialHeading,
+      // -89 to avoid discontinuity at -90 degrees pitch
+      CSMath.clamp(initialPitch - dyTotal * pitchFactor, CSMath.toRadians(-89), CSMath.toRadians(-15)),
```

Comment:
The max and min should be named constants so it's easier to find them to tweak them down the line.